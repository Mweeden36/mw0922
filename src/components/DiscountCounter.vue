<template>
  <div class="discount-amount__wrapper">
    <b class="checkout-form__label">Discount Amount:</b>
    <input
      v-model="discountAmount"
      type="number"
      class="checkout-discount"
      :class="invalid ? 'invalid' : ''"
      @keydown="onKeydown"
    />
    <div v-if="invalid" class="error-notification color--danger">
      Please make sure the discount amount is between 0 and 100%.
    </div>
  </div>
</template>

<script>
export default {
  name: "DiscountCounter",

  props: {
    modelValue: {
      type: Number,
      default: 0,
    },
  },

  data() {
    return {
      discountAmount: this.value,
      specialKeys: ["ArrowLeft", "ArrowRight", "Backspace", "Delete", "-", "+"],
    };
  },

  computed: {
    invalid() {
      return this.discountAmount > 100 || this.discountAmount < 0;
    },
  },

  methods: {
    onKeydown(e) {
      if (this.specialKeys.indexOf(e.key) === -1 && /[^0-9]/.test(e.key)) {
        // If it's not in the list of special keys, and not 0-9, don't
        // allow the user to type it.
        e.preventDefault();
      }
    },
  },

  watch: {
    discountAmount() {
      this.$emit("update:modelValue", this.discountAmount);
    },
  },
};
</script>

<style lang="scss" scoped>
.discount-amount__wrapper {
  .checkout-discount {
    outline: none;
    &.invalid {
      border-color: red;
    }
  }
}
</style>
