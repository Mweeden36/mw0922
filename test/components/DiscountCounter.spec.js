import DiscountCounter from "../../src/components/DiscountCounter.vue";
import { describe, it, expect, vi } from "vitest";
import { mount } from "@vue/test-utils";

describe("DiscountCounter.vue", () => {
  it("should Initialize with sensible defaults", () => {
    const { vm } = mount(DiscountCounter, {
      props: {
        modelValue: 0,
      },
    });
    expect(vm.discountAmount).toEqual(0);
    expect(vm.invalid).toEqual(false);
  });

  it("should generate the correct html", () => {
    const wrapper = mount(DiscountCounter, {
      props: {
        modelValue: 0,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should mark as invalid when discount amount is initialized as negative", () => {
    const { vm } = mount(DiscountCounter, {
      props: {
        modelValue: -1,
      },
    });
    expect(vm.discountAmount).toEqual(-1);
    expect(vm.invalid).toEqual(true);
  });

  it("should mark as invalid when discount amount is initialized as over 100", () => {
    const { vm } = mount(DiscountCounter, {
      props: {
        modelValue: 101,
      },
    });
    expect(vm.discountAmount).toEqual(101);
    expect(vm.invalid).toEqual(true);
  });

  it("should mark as invalid when discount amount is modified to over 100", () => {
    const { vm } = mount(DiscountCounter, {
      props: {
        modelValue: 100,
      },
    });

    expect(vm.discountAmount).toEqual(100);
    expect(vm.invalid).toEqual(false);
    vm.discountAmount = 101;
    expect(vm.invalid).toEqual(true);
  });

  it("should prevent alpha keys and allow numeric keys in the input", () => {
    const { vm } = mount(DiscountCounter, {
      props: {
        modelValue: 100,
      },
    });

    const spy = vi.fn();
    vm.onKeydown({ key: "a", preventDefault: spy });
    expect(spy).toHaveBeenCalledTimes(1);
    vm.onKeydown({ key: "0", preventDefault: spy });
    expect(spy).toHaveBeenCalledTimes(1);
    vm.onKeydown({ key: ".", preventDefault: spy });
    expect(spy).toHaveBeenCalledTimes(2);
  });
});
