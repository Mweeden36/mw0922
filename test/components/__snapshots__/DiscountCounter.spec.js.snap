// Vitest Snapshot v1

exports[`DiscountCounter.vue > should generate the correct html 1`] = `
"<div class=\\"discount-amount__wrapper\\" data-v-1eab4cab=\\"\\"><b class=\\"checkout-form__label\\" data-v-1eab4cab=\\"\\">Discount Amount:</b><input type=\\"number\\" class=\\"checkout-discount\\" data-v-1eab4cab=\\"\\">
  <!--v-if-->
</div>"
`;
